src/css3.rs

```diff
@@ -1,120 +1,959 @@
-//!
-//! Any
-//! :   [   IDENT
-//!     |   NUMBER
-//!     |   PERCENTAGE
-//!     |   DIMENSION
-//!     |   STRING
-//!     |   DELIM
-//!     |   URI
-//!     |   HASH
-//!     |   UNICODE-RANGE
-//!     |   INCLUDES
-//!     |   DASHMATCH
-//!     |   FUNCTION Any* RPAREN
-//!     |   LPAREN Any* RPAREN
-//!     |   LBRACKET Any* RBRACKET
-//!     ]
-//! ;
-//! ```
-//!
-pub mod new_parser;
-pub mod new_tokenizer;
-pub mod node;
-pub mod parser;
-pub mod tokenizer;
-pub mod tokens;
-pub mod unicode;
+mod new_tokenizer;
+mod unicode;
+mod nom_selector;
+mod ast;
+
+use crate::css3::new_tokenizer::{Token, TokenStreamer};
+use crate::css3::Error::{Syntax, UnexpectedEof};
+use core::fmt::Debug;
+use std::collections::HashMap;
+use log::{debug, trace};
+use thiserror::Error;
+use crate::css3::ast::Node;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("syntax error: {0}")]
+    Syntax(String),
+    #[error("unexpected end of stream")]
+    UnexpectedEof,
+}
+
+// =================================================================================================
+// Tokenstream is a simple already-tokenized stream of tokens. This is used for example when
+// parsing a declaration, which is a list of tokens that are already known.
+struct TokenStream {
+    tokens: Vec<Token>,
+    index: usize,
+}
+
+impl TokenStream {
+    fn new(tokens: Vec<Token>) -> TokenStream {
+        TokenStream { tokens, index: 0 }
+    }
+
+    fn new_from_componentvalues(values: Vec<ComponentValue>) -> TokenStream {
+        let tokens = values.iter().filter(|v| {
+            if let ComponentValue::PreservedToken(_) = v {
+                true
+            } else {
+                false
+            }
+        }).map(|v| {
+            if let ComponentValue::PreservedToken(token) = v {
+                token.clone()
+            } else {
+                panic!("we should not be here");
+            }
+        }).collect::<Vec<Token>>().into();
+
+        TokenStream{
+            tokens,
+            index: 0
+        }
+    }
+}
+
+impl TokenStreamer for TokenStream {
+    fn current(&self) -> Token {
+        if self.index >= self.tokens.len() {
+            return Token::Eof;
+        }
+
+        self.tokens[self.index].clone()
+    }
+
+    fn lookahead(&self, n: usize) -> Token {
+        if self.index + n >= self.tokens.len() {
+            return Token::Eof;
+        }
+
+        self.tokens[self.index + n].clone()
+    }
+
+    fn consume(&mut self) -> Token {
+        if self.index >= self.tokens.len() {
+            return Token::Eof;
+        }
+
+        let token = self.tokens[self.index].clone();
+        self.index += 1;
+        token
+    }
+
+    fn reconsume(&mut self) {
+        if self.index > 0 {
+            self.index -= 1;
+        }
+    }
+}
+
+// =================================================================================================
+
+pub struct CSS3ParserTng<'a> {
+    token_stream: &'a mut dyn TokenStreamer,
```

Comment:
When i think about this, it's probably better to use Nico's suggestion. (with the lifetime in the `Box`) because then you can store the `TokenStreamer` actually in the struct. If it doesn't matters, you can leave it as it is.